src/pal/src/misc/sysinfo.cpp

```diff
@@ -525,5 +525,20 @@ PAL_GetLogicalProcessorCacheSizeFromOS()
     }
 #endif
 
+#ifdef HAVE_SYSCTL
+    if (cacheSize == 0)
+    {
+        size_t cacheSizeFromSysctl = 0;
+        size_t sz = sizeof(cacheSizeFromSysctl);
+        const bool success = sysctlbyname("hw.l3cachesize", &cacheSizeFromSysctl, &sz, nullptr, 0) == 0
+            || sysctlbyname("hw.l2cachesize", &cacheSizeFromSysctl, &sz, nullptr, 0) == 0
+            || sysctlbyname("hw.l1dcachesize", &cacheSizeFromSysctl, &sz, nullptr, 0) == 0;
+        if (success)
+        {
+            cacheSize = cacheSizeFromSysctl;
+        }
+    }
+#endif
+
```

Comment:
Would it make sense to have an assert that cacheSize > 0, and then if it still is 0 return some default > 0?

Or should the default computation fallback based on PAL_GetLogicalCpuCountFromOS above instead be moved here and made unconditional?

Reply:
It looks like this function is only used by `GetCacheSizePerLogicalCpu`, which is only used by gc, which manually already provides its own default each time it calls that. It might indeed clean up the code a bit to move all those defaults here.
I assume you're referring to `cacheSize = logicalCPUs*std::min(1536, std::max(256, (int)logicalCPUs*128))*1024;`? I guess it doesn't make sense to have two separate locations for default cache sizes (here vs in gc), though it might not make sense to have the same default on different systems.

Reply:
The semantics of the GetLogicalProcessorCacheSizeFromOS is to return 0 when it cannot determine the cache size. The PAL_GetLogicalProcessorCacheSizeFromOS is a unix version of that function and thus its implementation followed the same logic.
https://github.com/dotnet/coreclr/blob/580152a50e4092d32c7d05ec876cd6976483ebf1/src/vm/util.cpp#L1461-L1463.
I wonder if GC should be fixed too to handle this case in a more sensible way instead of using so small gen0 size.